keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -398,64 +447,216 @@ which would have the following behavior, both for CRD types and built-in types:
+wire. CRD defaulting will not default a zero value though.
 
-Both these cases will be handled properly, here's an example:
 ```golang
 type Object struct {
-     // This field has an invalid zero-value, empty string is not a
-     // valid string.
-     // +default=default-name
-     Name string
+     // This field is omit-empty, and hence Golang unmarshalling drops a zero-value.
+     // +default="default-name"
+     Name string `json:"name,omitempty"`
 
-     // This field is defaulted to 0 with no defaulter.
+     // Non-pointer scalar fields without omitempty require zero-value default.
      // +default=0
      Defaulted int
 }
 ```
 
-This would generate the following OpenAPI:
+Non-pointer scalar fields that have a default must be omitempty (unless
+their default is the zero-value). Kubebuilder and built-in defaulting
+generators should both require this. Our type above fulfills this
+requirement, in contrast to
+
+```golang
+type Invalid struct {
+     // This field is NOT omit-empty, and CRD and built-in defaulting would differ,
+     // even when going through Golang types on the client
+     // +default="default-name"
+     Name string `json:"name"`
+}
+```
+
+The valid Object type would generate the following OpenAPI:
 ```yaml
 Object:
   type: object
   properties:
-    invalid:
+    name:
       type: string
-      default: default-name
+      default: "default-name"
     defaulted:
-      type: boolean
-      default: false
+      default: 0
+      type: integer
 ```
 
 Which, for both CRD types and built-in types, would generate the following behavior:
 ```python
 >>> default('{}')
 {"entry": {"name": "default-name", "number": 0}}
->>> default('{entry: {name: other-name}}')
-{"entry": {"name": "other-name", "number": 0}}
+>>> default('{name: other-name}')
+{"name": "other-name", "defaulted": 0}
+# For built-in, additionally, we would have:
+>>> default('{name: ""}')
+{"name": "default-name", "defaulted": 0}
 ```
 
-Note that the generated code for built-in will look like this:
+Note that a simplified version, which deserializes on every call, of the generated code for
+built-ins would look like this:
 ```golang
 func GenerateDefault(obj *Object) {
      if obj.Name == "" {
-          obj.Name = "default-name"
+          if err := json.Unmarshal([]byte(`"default-name"`), &obj.Name); err != nil {
+               panic(err)
+          }
      }
      if obj.Number == 0 {
-          obj.Number = 0
+          if err := json.Unmarshal([]byte(`0`), &obj.Number); err != nil {
+               panic(err)
+          }
      }
 }
 ```
 
+It's still possible to specify an invalid zero-value in built-in types
+embedded into a CRD, but client-go will drop these values, thanks to
+`omitempty`. On the other hand, a manifest in JSON
+
+{"entry": {"invalid": "", "number": 0}}
+
+sent to the server would result in different behaviour for built-ins
+
+```python
+>>> default('{name: ""}')
```

Comment:
NIT: it's a bit weird that the example object changes when it looks like you're going to keep talking about the above invalid entry.